### ISSUE TITLE
FileSystem sync operations ?

### ISSUE THREAD

**Issue by WebReflection** [voodooattack/nexusjs#9]
I'm working on a CommonJS require functionality based on ESM module, example:
```js
import cjs from "module";
const require = cjs(import.meta.dirname);

const mod = require('some-module');
// require.cache
// require.resolve
```
and while I know JSC exposes `readFile` I am trying to target NexusJS too (beside browsers) but I cannot find anything like `readFileSync` or `statSync`

CommonJS is inevitably based on synchronous load so ... is there any plan to make it possible ?

**Comment by voodooattack** [comment]
It's currently possible to read files in a synchronous manner, although it's a bit unwieldy.

```js
const fileSource = new Nexus.IO.FilePullDevice(import.meta.filename);
const buffer = fileSource.readSync(1024);
const decoder = new Nexus.IO.UTF8StringFilter();
const text = decoder.processSync(buffer);

console.log(text);
```

I recommend waiting on the CJS work until I push the next commit, since I'm in the processes of implementing a ModuleFetcher system.

Fetchers will be responsible for loading the different module types, and work like this:

```js
import { ESM } from './file'; // regular esm module
import _ from 'npm:lodash'; // module loaded via the `npm` fetcher.
import MyCJSModule from 'cjs:~/devel/my-commonjs-module'; // a local `common.js` module
/* ... etc ... */
```

You'll be able to register fetchers at run-time. They are also inherited. So, let's say you import `npm:loadash`, any module required by `lodash` would then be loaded via the same `npm` fetcher, unless explicitly overridden.

I hope I'm making sense? I'm still in the process of fleshing this out. So I'm open to suggestions.

**Comment by voodooattack** [comment]
It may be worth clarifying a few things:

* Nexus will load dependencies at run-time, there is no `npm install` step for installing dependencies.
* All modules will be stored under `~/.nexus/cache/${fetcher}/${package}@${version}` in a flat directory structure.
* You can specify the desired package version in the `import` statement directly: `import x from 'x@~2.0.0';`
* A second source of truth regarding package versions is the nearest `package.json` file up the project hierarchy. (You can nest independent `package.json` files this way and override package versions for a specific subdirectory)

**Comment by WebReflection** [comment]
the standard direction to define ESM loaders is different:
https://github.com/whatwg/html/issues/3161

the moment you define your own way to import you're automatically opting out from cross modules compatibility, are you sure you want to go this way?

There have been months of discussions about CJS in ESM and _magic_ has never been the right answer while `explicit` intent seems to be the winner, and NodeJS is going that way too.

That means, you either exposes `import.meta.require` to require _CJS_ modules, or you explicitly include them via proper CJS mechanism, like the example I've shown before:

```js
import cjs from "module";
const require = cjs(import.meta.dirname);

const mod = require('some-module');
// require.cache
// require.resolve
```

this would keep separated CJS from ESM so that live bindings are not necessary, hacks on CJS are not necessary, and so on and so fort.

If you want to use `package.json` file you should also consider each module might have a `main` or a `module` field that points respectively at the _CJS_ or the _ESM_ version of the module so there's also less to guess about what to import, which would always be ESM by default.

Trust me when I say I've been around modules for months, and in contact with TC39 as well TSC NodeJS members about this ... proposing a diversion *now*, and at this stage, looks like a very unwise decision to me.

It's your project though, you do what you think it's best.

**Comment by voodooattack** [comment]
I was not aware of that discussion, I'll study more on the subject before committing to anything.

Thank you for bringing it to my attention.

**Comment by WebReflection** [comment]
more on the subject:
https://medium.com/the-node-js-collection/the-current-state-of-implementation-and-planning-for-esmodules-a4ecb2aac07a

**Comment by jtenner** [comment]
@voodooattack Please let me know if you need help. People who want to write their modules can transpile them using babel and take advantage of this syntax.

```
import cjs from "module";
const require = cjs(import.meta.dirname);

const mod = require('some-module');
```
I am happy to offer my time to write and maintain a babel plugin that does this.

Let me know when this feature is supported.

**Comment by WebReflection** [comment]
FWIW I've just published an ESM module that does that. It works already in both browsers and SpiderMonkey, NexusJS might be next (it takes forever to download WebKit, I have crappy i0nternet connection these days and it's impossible to do that!!!)

The module is called [cjs4esm](https://github.com/WebReflection/cjs4esm) :tada:

**Comment by voodooattack** [comment]
@WebReflection That's excellent. I've adapted the code and I plan to use the Node.js hook api in the future.

Is this a better looking solution?

```js
import _ from 'npm://lodash';
```

**Comment by WebReflection** [comment]
> Is this a better looking solution?

anything that uses ESM standard syntax to import incompatible CJS is a bad solution, IMO

**Comment by voodooattack** [comment]
@WebReflection Good point. I won't make it interop with CJS. Merely import from `npm` as is. It'll be up to the developer to configure CJS-interop on their own.

**Comment by WebReflection** [comment]
> Merely import from npm as is

what does that mean in practice?

will you use the `package.json` `"module"` entry by default and throw if not available?

**Comment by WebReflection** [comment]
btw, `npm://` or others are protocol look like a better approach regardless

**Comment by voodooattack** [comment]
@WebReflection Yes, if `module` is not found in `package.json` it'll throw. Unless the user imports a specific path like this:
```js 
import package from 'npm://mypackage/dist/mymodule.esm';
```

I thought that specifying the package source in the protocol looked cleaner, at least it has a well-known structure that can be parsed by utilities easier.

**Comment by voodooattack** [comment]
@WebReflection Sorry for the inactivity.

I’m currently working on a big update. But no worries! You’re right to close this for now, because I’m most likely changing a lot of the things discussed here.
